validator/src/journal/block_manager.rs

```diff
@@ -0,0 +1,1039 @@
+            .get(block_id)
+            .map(|anchor| anchor.internal_ref_count)
+    }
+
+    fn increase_external_ref_count(&mut self, block_id: &str) {
+        if let Some(anchor) = self.anchors_by_block_id.get_mut(block_id) {
+            anchor.external_ref_count += 1;
+        }
+    }
+
+    fn decrease_external_ref_count(&mut self, block_id: &str) {
+        if let Some(anchor) = self.anchors_by_block_id.get_mut(block_id) {
+            anchor.external_ref_count -= 1;
+        }
+    }
+
+    fn get_external_ref_count(&self, block_id: &str) -> Option<i64> {
+        self.anchors_by_block_id
+            .get(block_id)
+            .map(|anchor| anchor.external_ref_count)
+    }
+}
+
+struct Anchor {
+    pub blockstore_name: String,
+    pub external_ref_count: i64,
+    pub internal_ref_count: i64,
+    pub block_num: u64,
+    pub block_id: String,
+}
+
+#[derive(Default)]
+pub struct BlockManager {
+    block_by_block_id: HashMap<String, (Block, i64, i64)>,
+
+    blockstore_by_name: HashMap<String, Box<BlockStore>>,
+
+    anchors: Anchors,
+}
+
+/// The BlockManager maintains integrity of all the blocks it contains,
+/// such that for any Block within the BlockManager,
+/// that Block's predecessor is also within the BlockManager.
+impl BlockManager {
+    pub fn new() -> Self {
+        BlockManager::default()
+    }
+
+    fn contains(&self, block_id: &str) -> bool {
+        let in_memory = self.block_by_block_id.contains_key(block_id);
+
+        let in_any_blockstore = self.blockstore_by_name
+            .values()
+            .any(|blockstore| blockstore.get(vec![block_id.into()]).count() > 0);
+        let is_root = block_id == NULL_BLOCK_IDENTIFIER;
+
+        in_memory || in_any_blockstore || is_root
+    }
+
+    /// Put is idempotent, making the guarantee that after put is called with a
+    /// block in the vector argument, that block is in the BlockManager
+    /// whether or not it was already in the BlockManager.
+    /// Put makes three other guarantees
+    ///     - If the zeroth block in branch does not have its predecessor
+    ///       in the BlockManager an error is returned
+    ///     - If any block after the zeroth block in branch
+    ///       does not have its predecessor as the block to its left in
+    ///       branch, an error is returned.
+    ///     - If branch is empty, an error is returned
+    pub fn put(&mut self, branch: Vec<Block>) -> Result<(), BlockManagerError> {
+        match branch.split_first() {
+            Some((head, tail)) => {
+                if !self.contains(head.previous_block_id.as_str()) {
+                    return Err(BlockManagerError::MissingPredecessor(format!(
+                        "During Put, missing predecessor of block {}: {}",
+                        head.header_signature, head.previous_block_id
+                    )));
+                } else {
+                    let mut previous = None;
+
+                    for block in tail {
```

Comment:
This internal branch integrity check feels like a good candidate for refactoring into a separate private helper function with a descriptive name so it is easier to tell what is going on.